### ISSUE TITLE
Feature: Property that checks that a function is called n times

### ISSUE THREAD

**Issue by HugoPeters1024** [nomeata/inspection-testing#70]
When creating fusion rules like

```haskell
{-# Rules
   "mapTree/mapTree"   forall f g.   mapTree f . mapTree g    = mapTree (f. g)   ;
#-}
```

It can be valuable to assert that certain functions are reduced correctly by checking that say `mapTree` occurs exactly once in the core. (I realise the example is somewhat contrived because implementing `Functor` for `Tree a` gives us access to all the `fmap` fusion rules but I think my point this stands)

**Comment by phadej** [comment]
I think this is general non-trivial. Total absence is easy.

```haskell
let m = mapTree in \x -> m f (m g x)
``` 

How many times is `mapTree` called?

Or if GHC decides to push (an optimized!) expression in to `case` branches:

```haskell
case e of
   Left x -> mapTree (f . g) ...
   Right x -> mapTree (f . g) ...
```

**Comment by HugoPeters1024** [comment]
That's a valid concern! I agree that I don't think a useful solution really exists. For the latter case you could try each code path but that increases work exponentially. For the first case you would have to make some analysis whether a variable is 'tainted' with another which leads to many false positives (positives meaning that the test will fail). 

I'm currently working on my own core inspection tool where I've discovered that detecting whether rewrite rules have fired in a specific expression is to my knowledge is near impossible to detect (especially if multiple overlapping rules fire), otherwise that would be an amazing property to have. 

Something like:

```haskell
inspect $ 'f (RuleFired "mapTree/mapTree")
```

Thanks for your thoughts nonetheless, closing.
